File: sdk/javascript/src/vector.ts
import type { CloudPool } from './client';
import type { VectorCollection, VectorDocument, VectorSearchResult } from './types';

export class VectorClient {
  constructor(private readonly cp: CloudPool) {}

  /** List all vector collections */
  async listCollections(): Promise<VectorCollection[]> {
    return this.cp.request<VectorCollection[]>('GET', '/api/vector/collections');
  }

  /** Create a new vector collection */
  async createCollection(
    name: string,
    description: string,
    dimension = 1536,
    distanceMetric = 'cosine',
  ): Promise<VectorCollection> {
    return this.cp.request<VectorCollection>('POST', '/api/vector/collections', {
      name,
      description,
      dimension,
      distanceMetric,
    });
  }

  /** Delete a collection by ID */
  async deleteCollection(collectionId: string): Promise<void> {
    await this.cp.request('DELETE', `/api/vector/collections/${collectionId}`);
  }

  /**
   * Index a document into a collection.
   * The SDK sends the content; CloudPool generates the embedding automatically.
   */
  async indexDocument(
    collectionId: string,
    docId: string,
    content: string,
    metadata?: Record<string, unknown>,
  ): Promise<VectorDocument> {
    return this.cp.request<VectorDocument>('POST', `/api/vector/collections/${collectionId}/documents`, {
      docId,
      content,
      metadata,
    });
  }

  /** Semantic search within a specific collection */
  async searchCollection(
    collectionId: string,
    query: string,
    limit = 10,
  ): Promise<VectorSearchResult[]> {
    return this.cp.request<VectorSearchResult[]>(
      'GET',
      `/api/vector/collections/${collectionId}/search?query=${encodeURIComponent(query)}&limit=${limit}`,
    );
  }

  /** Semantic search across all user files */
  async searchFiles(query: string): Promise<VectorSearchResult[]> {
    return this.cp.request<VectorSearchResult[]>(
      'GET',
      `/api/vector/search?query=${encodeURIComponent(query)}`,
    );
  }
}

File: sdk/python/cloudpool/vector.py
from __future__ import annotations

from typing import Any


class VectorClient:
    """Vector search APIs."""

    def __init__(self, cp):
        self.cp = cp

    def list_collections(self) -> list[dict[str, Any]]:
        return self.cp.request(
            "GET",
            "/api/vector/collections",
        )

    def create_collection(
        self,
        name: str,
        description: str,
        dimension: int = 1536,
        distance_metric: str = "cosine",
    ) -> dict[str, Any]:
        return self.cp.request(
            "POST",
            "/api/vector/collections",
            json={
                "name": name,
                "description": description,
                "dimension": dimension,
                "distanceMetric": distance_metric,
            },
        )

    def delete_collection(
        self,
        collection_id: str,
    ) -> None:
        self.cp.request(
            "DELETE",
            f"/api/vector/collections/{collection_id}",
        )

    def index_document(
        self,
        collection_id: str,
        doc_id: str,
        content: str,
        metadata: dict[str, Any] | None = None,
    ) -> dict[str, Any]:
        return self.cp.request(
            "POST",
            f"/api/vector/collections/{collection_id}/documents",
            json={
                "docId": doc_id,
                "content": content,
                "metadata": metadata,
            },
        )

    def search_collection(
        self,
        collection_id: str,
        query: str,
        limit: int = 10,
    ) -> list[dict[str, Any]]:
        return self.cp.request(
            "GET",
            f"/api/vector/collections/{collection_id}/search?query={query}&limit={limit}",
        )

    def search_files(
        self,
        query: str,
    ) -> list[dict[str, Any]]:
        return self.cp.request(
            "GET",
            f"/api/vector/search?query={query}",
        )